jerry-ext/module/module.c

```diff
@@ -0,0 +1,218 @@
+                            jerry_value_t module_name, /**< key at which to cache the module */
+                            jerry_value_t module) /**< the module to cache */
+{
+  jerry_value_t ret = jerry_set_property (cache, module_name, module);
+
+  if (jerry_value_has_error_flag (ret))
+  {
+    jerry_release_value (module);
+  }
+  else
+  {
+    jerry_release_value (ret);
+    ret = module;
+  }
+
+  return ret;
+} /* jerryx_module_add_to_cache */
+
+/**
+ * Declare and define the default module resolver - one which examines what modules are defined in the above linker
+ * section and loads one that matches the requested name, caching the result for subsequent requests using the context
+ * data mechanism.
+ */
+JERRYX_MODULE_RESOLVER (native_resolver)
+{
+  int index;
+  const jerryx_module_t *module_p = NULL;
+  jerry_value_t ret = 0;
+
+  /* Look for the module by its name in the list of module definitions. */
+  for (index = 0, module_p = &__start_jxm_modules[0];
+  &__start_jxm_modules[index] < __stop_jxm_modules;
+  index++, module_p = &__start_jxm_modules[index])
+  {
+    if (!strcmp ((char *) module_p->name, (char *) name))
+    {
+      /* If we find the module by its name we load it and cache it if it has an on_resolve () and complain otherwise. */
+      ret = (module_p->on_resolve) ? module_p->on_resolve () :
+      jerryx_module_create_error (JERRY_ERROR_TYPE, on_resolve_absent, name);
+      break;
+    }
+  }
+
+  return ret;
+} /* JERRYX_MODULE_RESOLVER */
+
+/**
+ * Declare the section where we expect to find module resolvers - at least the one above will be there.
+ */
+typedef jerry_value_t (*jerryx_module_resolver_t)(const jerry_char_t *);
+JERRYX_SECTION_DECLARE (jxm_resolvers, jerryx_module_resolver_t)
+
+/**
+ * Resolve a single module using the module resolvers available in the section declared above and load it into the
+ * current context.
+ *
+ * @return a jerry_value_t containing one of the followings:
+ *   - the result of having loaded the module named @p name, or
+ *   - the result of a previous successful load, or
+ *   - an error indicating that something went wrong during the attempt to load the module.
+ */
+jerry_value_t
+jerryx_module_resolve (const jerry_char_t *name) /**< name of the module to load */
+{
+  int index;
+  const jerryx_module_resolver_t *resolver_p;
+  jerry_value_t ret;
+  jerry_value_t instances = *(jerry_value_t *) jerry_get_context_data (&jerryx_module_manager);
+  jerry_value_t module_name = jerry_create_string_from_utf8 (name);
+
+  /* Return the cached instance if present. */
+  ret = jerryx_module_check_cache (instances, module_name);
+  if (ret != 0)
+  {
+    goto done;
+  }
+
+  /* Try each resolver in the section until one manages to return non-zero. */
+  for (index = 0, resolver_p = &__start_jxm_resolvers[0];
+  &__start_jxm_resolvers[index] < __stop_jxm_resolvers;
+  index++, resolver_p = &__start_jxm_resolvers[index])
+  {
```

Comment:
Why do we need multiple resolver-functions? And How can user make sure the order those functions?

I think one resolver function is enough, and user can set its own one if they like.